### DISCUSSION TITLE
Deploy Yourls via GitHub Actions - how to 'inject' secrets into config.php?

### DISCUSSION THREAD

**Discussion by jfix** [YOURLS/YOURLS#2962] [Q&A] [answered]
Hi 👋

I'd like to start using GitHub Actions to deploy our URL shortener instance. In itself this seems relatively straight-forward except for the handling of certain sensitive data, especially in the `config.php` file (database login and password, for example, but also an array of authorized IP addresses used by a plugin of ours, etc.).

Of course, I can add those strings to the project's secrets, but then I wonder what the "best practice" is for getting them into the `config.php` file. 

Also, the array of authorized domain names (for which we permit the creation of short URLs) is updated relatively frequently, so the secrets section is maybe not the best place.

For the credentials, I guess substituting placeholder tokens is a valid way. I found this approach on Stack Overflow:
https://stackoverflow.com/questions/66724889/github-actions-substitute-string-before-deploying

Not being a PHP pro, I'm not sure whether that's the way to go, or whether there is another, more appropriate way.  As I've seen that the YOURLS project itself is using GitHub Actions, I am hoping someone can recommend a way to do.

Grateful for your input!

**Comment by ozh** [comment, marked as answer]
The way described in this stackoverflow answer would totally work. Make your `config.php` with `define( 'YOURLS_DB_USER', '__DBUSERNAME__' );` and use the same script with sed to replace `__DBUSERNAME__` with `secrets.DBUSERNAME`

**Reply by jfix** [reply]
Thanks @ozh, I'll take that as a confirmation of a "best practice". I was kind of hoping that PHP has something equivalent to node's `process.env` and `.env` files, but as I couldn't find anything PHP-generic (i.e. not Symphony or Deployer-related), I guess that will be the way I go. Thanks again.

PS: I may have not looked good enough the first time around, because apparently `.env` is a thing in PHP-land, too: https://www.codementor.io/@ccornutt/keeping-credentials-secure-in-php-kvcbrk55z. Would that be worth suggesting for a future release? ;-)

**Reply by ozh** [reply]
More of a "that would definitely work and it's not stupid" than a "best practice". Often, [there is more than one way to do it](https://en.wikipedia.org/wiki/There%27s_more_than_one_way_to_do_it) and several ways are as good as another.

PHP dotenv would be another very valid approach : modify your config.php to load the dotenv lib and define your variables from a custom .env file.

In any case it might be a good idea to disable the automatic password encryption in your config file, see [here](https://github.com/YOURLS/YOURLS/wiki/Username-Passwords#but-i-dont-want-to-encrypt-my-password-)

**Reply by jfix** [reply]
Just to provide some closure to the subject, here is the code that does what I want (not yet integrated in YOURLS) using `phpdotenv` as discussed:

Somewhere in the GitHub Actions job, you have to create the `.env` file inside the GitHub Actions workspace, based on the secrets stored in GitHub:

```
    - run: |
        echo "A=${{ secrets.A }}" > .env
        echo "B=${{ secrets.B }}" >> .env
        echo "C=${{ secrets.C }}" >> .env
```

Later on, you need to transfer this file to the correct remote location:

```
    - name: Copy .env to remote server
      uses: Creepios/sftp-action@v1.0.1
      with:
        host: ${{ env.REMOTE_HOST }}
        port: 22
        username: ${{ env.REMOTE_USER }}
        privateKey: ${{ secrets.SSH_PRIVATE_KEY }}
        localPath: '.env'
        remotePath: './${{ env.REMOTE_PATH }}/.env'
```

In the PHP code, you can then reference the environment variables like so:

```lang-php
<?php
    require_once __DIR__.'/../vendor/autoload.php';
    $dotenv = Dotenv\Dotenv::createImmutable(__DIR__.'/../');
    $dotenv->load();
    $dotenv->required(['A', 'B', 'C']);
?>
<html>
<h1>Environment variable test</h1>
<br/> A <?php var_dump( $_ENV['A'] ); ?>
<br/> B <?php var_dump( $_ENV['B'] ); ?>
<br/> C <?php var_dump( $_ENV['C'] ); ?>
</html>
```
